generator/intermediate_data.cpp

```diff
@@ -0,0 +1,250 @@
+
+void IndexFileWriter::WriteAll()
+{
+  if (m_elements.empty())
+    return;
+
+  m_fileWriter.Write(&m_elements[0], m_elements.size() * sizeof(Element));
+  m_elements.clear();
+}
+
+void IndexFileWriter::Add(Key k, Value const & v)
+{
+  if (m_elements.size() > kFlushCount)
+    WriteAll();
+
+  m_elements.emplace_back(k, v);
+}
+
+// OSMElementCacheReader ---------------------------------------------------------------------------
+OSMElementCacheReader::OSMElementCacheReader(string const & name, bool preload)
+  : m_fileReader(name), m_offsets(name + OFFSET_EXT), m_name(name), m_preload(preload)
+{
+  if (!m_preload)
+    return;
+  size_t sz = m_fileReader.Size();
+  m_data.resize(sz);
+  m_fileReader.Read(0, m_data.data(), sz);
+}
+
+void OSMElementCacheReader::LoadOffsets() { m_offsets.ReadAll(); }
+
+// OSMElementCacheWriter ---------------------------------------------------------------------------
+OSMElementCacheWriter::OSMElementCacheWriter(string const & name, bool preload)
+  : m_fileWriter(name), m_offsets(name + OFFSET_EXT), m_name(name), m_preload(preload)
+{
+}
+
+void OSMElementCacheWriter::SaveOffsets() { m_offsets.WriteAll(); }
+
+// RawFilePointStorageMmapReader -------------------------------------------------------------------
+RawFilePointStorageMmapReader::RawFilePointStorageMmapReader(string const & name)
+  : m_mmapReader(name)
+{
+}
+
+bool RawFilePointStorageMmapReader::GetPoint(uint64_t id, double & lat, double & lon) const
+{
+  LatLon ll;
+  m_mmapReader.Read(id * sizeof(ll), &ll, sizeof(ll));
+
+  bool ret = FromLatLon(ll, lat, lon);
+  if (!ret)
+    LOG(LERROR, ("Node with id =", id, " not found!"));
+  return ret;
+}
+
+// RawFilePointStorageWriter -----------------------------------------------------------------------
+RawFilePointStorageWriter::RawFilePointStorageWriter(string const & name) : m_fileWriter(name) {}
+
+void RawFilePointStorageWriter::AddPoint(uint64_t id, double lat, double lon)
+{
+  LatLon ll;
+  ToLatLon(lat, lon, ll);
+
+  m_fileWriter.Seek(id * sizeof(ll));
+  m_fileWriter.Write(&ll, sizeof(ll));
+
+  ++m_numProcessedPoints;
+}
+
+// RawMemPointStorageReader ------------------------------------------------------------------------
+RawMemPointStorageReader::RawMemPointStorageReader(string const & name)
+  : m_fileReader(name), m_data(static_cast<size_t>(1) << 33)
+{
+  m_fileReader.Read(0, m_data.data(), m_data.size() * sizeof(LatLon));
+}
+
+bool RawMemPointStorageReader::GetPoint(uint64_t id, double & lat, double & lon) const
+{
+  LatLon const & ll = m_data[id];
+  bool ret = FromLatLon(ll, lat, lon);
+  if (!ret)
+    LOG(LERROR, ("Node with id =", id, " not found!"));
+  return ret;
+}
+
+// RawMemPointStorageWriter ------------------------------------------------------------------------
+RawMemPointStorageWriter::RawMemPointStorageWriter(string const & name)
+  : m_fileWriter(name), m_data(static_cast<size_t>(1) << 33)
```

Comment:
Could you please explain the meaning of the constant: `static_cast<size_t>(1) << 33`? Some big (unreachable) number? Why not `numeric_limits<size_t>::max()`?

Reply:
As far as I understand, we need a large enough constant-size buffer and do not want to calculate its size exactly.
It is intended to be nearly reachable, 2^64 is too much to allocate.

Reply:
static const size_t kMaxPointsInOSM = static_cast<size_t>(1) << 33;

Reply:
done